File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    public static void gp(int n, int left, int right, String s, List<String> list){
        if(s.length()==2*n){
            list.add(s);
            return;
        }
        if(left<n) gp(n,left+1,right,s+"(",list);
        if(right<left) gp(n,left,right+1,s+")",list);
    }
    public List<String> generateParenthesis(int n) {
        
        List<String> list = new ArrayList<>();
        String s = "";
        gp(n,0,0,s,list);
        return list;
    }
}

File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
    void rug(int close, int open, string s, int n, vector<string>& result) {
        if (close + open >= n * 2) {
            result.push_back(s);
            return;
        }
        if (close > open)
            return;
        if (open == n) {
            rug(close + 1, open, s + ')', n, result);
        } else {
            rug(close, open + 1, s + '(', n, result);
            rug(close + 1, open, s + ')', n, result);
        }
    }

    vector<string> generateParenthesis(int n) {
        vector<string> result;
        rug(0, 0, "", n, result);
        return result;
    }
};